pkg/ingester/wal.go

```diff
@@ -156,12 +156,21 @@ func (w *walWrapper) run() {
 		return
 	}
 
-	ticker := time.NewTicker(w.cfg.CheckpointDuration)
-	defer ticker.Stop()
-
+	lastCheckpoint := time.Now().Add(-w.cfg.CheckpointDuration)
 	for {
 		select {
-		case <-ticker.C:
+		case <-w.quit:
+			level.Info(util.Logger).Log("msg", "creating checkpoint before shutdown")
+			if err := w.performCheckpoint(true); err != nil {
+				level.Error(util.Logger).Log("msg", "error checkpointing series during shutdown", "err", err)
+			}
+			return
+		default:
+			timeElapsedSinceLastCheckpoint := time.Now().Sub(lastCheckpoint)
+			if timeElapsedSinceLastCheckpoint < w.cfg.CheckpointDuration {
+				time.Sleep(w.cfg.CheckpointDuration - timeElapsedSinceLastCheckpoint)
+			}
```

Comment:
This is better than the ticker because: we do not take into account the time taken to write the checkpoint and new series added in between. So checkpoint usually takes more than the checkpoint duration to write. This can cause a tick to be buffered up eventually if many checkpoints take more time. Hence if suddenly there is a drop in series and the checkpoint gets over before the checkpoint duration, the next checkpoint will be started before the checkpoint duration completition. This makes sure it never happens. (Microoptimization? Yes :P but checkpoints cause some disk throttling so lets avoid it)